### PR TITLE
Updates the state learning tutorial for TensorFlow 2.0

### PR DESCRIPTION
**Pull request by josh146**
**Context:** The existing TensorFlow tutorial was out of date (requiring the old TF1.3 backend), and in the form of a Jupyter notebook.

**Description of the Change:**

* Ports the tutorial to Sphinx-Gallery, by using the [converter script available here](https://gist.github.com/chsasank/7218ca16f8d022e02a9c0deb94a310fe).

* Ports the tutorial to use the new TF2 backend

**Benefits:** As the tutorial has been ported to Sphinx gallery, it is now executable, making it easy to test to see whether it provides the correct output.

**Possible Drawbacks:** Had to reduce the number of resources and optimization steps, since this tutorial will now be executed alongside the docs.

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by josh146** [comment]
[Rendered screenshot here](https://user-images.githubusercontent.com/2959003/80373138-bf77d500-88d3-11ea-9299-cbb06242de10.png)

File: examples_apps/run_state_learner.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+passive_sd = 0.1
+active_sd = 0.001
+
+######################################################################
+# The layer parameters :math:`\vec{\theta}`
+# -----------------------------------------
+#
+# We use TensorFlow to create the variables corresponding to the gate
+# parameters. Note that each variable has shape ``[depth]``, with each
+# individual element representing the gate parameter in layer :math:`i`.
```
Comment (trbromley) [inline examples_apps/run_state_learner.py:123]:
```suggestion
# parameters. Note that we focus on a single mode circuit where
# each variable has shape ``[depth]``, with each
# individual element representing the gate parameter in layer :math:`i`.
```
I was confused at first why this was the shape

File: examples_apps/run_state_learner.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+# Single-mode Strawberry Fields program
+prog = sf.Program(1)
+
+# Create the Strawberry Fields free parameters for each layer
+sf_params = []
+names = ["r1", "sq_r", "sq_phi", "r2", "d_r", "d_phi", "kappa"]
+
+for k in range(depth):
+    sf_params_names = ["{}_{}".format(n, k) for n in names]
+    sf_params.append(prog.params(*sf_params_names))
```
Comment (trbromley) [inline examples_apps/run_state_learner.py:163]: (very minor)
```suggestion
for i in range(depth):
    sf_params_names = ["{}_{}".format(n, i) for n in names]
    sf_params.append(prog.params(*sf_params_names))
```
Since in the next line we talk about the i-th layer.

File: examples_apps/run_state_learner.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+######################################################################
+# Now, we can create a function to define the :math:`i`\ th layer, acting
+# on qumode ``q``. This allows us to simply call this function in a loop
+# later on when we build our circuit.
```
Comment (trbromley) [inline examples_apps/run_state_learner.py:168]:
```suggestion
# on qumode ``q``. We add the :class:`~.utils.operation` decorator so that the layer can be used
# as a single operation when constructing our circuit within the usual
# Strawberry Fields Program context
```

File: examples_apps/run_state_learner.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+
+
+def cost(weights):
+    # Mapping from the Strawberry Fields free parameters to
+    # the TensorFlow variables.
+    mapping = {
+        p.name: w
+        for (sf_layer, weights_layer) in zip(sf_params, weights * 1.0)
+        for (p, w) in zip(sf_layer, weights_layer)
+    }
```
Comment (trbromley) [inline examples_apps/run_state_learner.py:246]: I find this part quite hard to parse. It would be good to have a separate line explaining this and maybe printing out the dictionary when applied to `weights`?

File: examples_apps/run_state_learner.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+    fidelity = tf.abs(tf.reduce_sum(tf.math.conj(ket) * target_state)) ** 2
+
+    # Objective function to minimize
+    cost = tf.abs(tf.reduce_sum(tf.math.conj(ket) * target_state) - 1)
+    return cost, fidelity, ket
+
+
+######################################################################
+# Now that the cost function is defined, we can define and run the
+# optimization as well as the quantum engine. Below, we choose the Adam
```
Comment (trbromley) [inline examples_apps/run_state_learner.py:265]:
```suggestion
# Now that the cost function is defined, we can define and run the
# optimization. Below, we choose the Adam
```
I felt this part was a bit technical and explained by the following sentence.

File: examples_apps/run_state_learner.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+
+    # one repitition of the optimization
+    gradients = tape.gradient(loss, weights)
+    opt.apply_gradients(zip([gradients], [weights]))
+
+    # Prints progress at every rep
+    if i % 1 == 0:
+        print("Rep: {} Cost: {:.4f} Fidelity: {:.4f}".format(i, loss, fid))
+
+
```
Comment (trbromley) [inline examples_apps/run_state_learner.py:304]: image

On my local version it didn't seem to train :thinking:

Comment (trbromley) [reply examples_apps/run_state_learner.py:304]: Which has a knock on effect to the rest of the results section

File: examples_apps/run_state_learner.py [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+opt = tf.keras.optimizers.Adam(learning_rate=lr)
+eng = sf.Engine("tf", backend_options={"cutoff_dim": cutoff})
+
+######################################################################
+# We then loop over all repetitions, storing the best predicted fidelity
+# value.
+
+fid_progress = []
+best_fid = 0
+
```
Comment (thisac) [inline examples_apps/run_state_learner.py:278]:
```suggestion
```

File: examples_apps/run_state_learner.py [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+   optimization landscape.
+
+-  ``reps``: the number of steps in the optimization routine performing
+   gradient descent
+
+Some other optional hyperparameters include:
+
+-  The standard deviation of initial parameters. Note that we make a
+   distinction between the standard deviation of *passive* parameters
+   (those that preserve photon number when changed, such as phase
```
Comment (antalszava) [inline examples_apps/run_state_learner.py:92]: (Minor suggestion)
```suggestion
   (those that preserve the number of photons when changed, such as phase
```

File: examples_apps/run_state_learner.py (hunk 1 above)
Comment (antalszava) [reply examples_apps/run_state_learner.py:123]: (Minor suggestion): how about having a tuple based shape representation for `depth`: ``(depth,)``? (gets converted internally to a tuple anyways, just thought otherwise questions might be raised why it is a `list` here)

File: examples_apps/run_state_learner.py (hunk 4 above)
Comment (antalszava) [reply examples_apps/run_state_learner.py:246]: Agreed! Was `weights * 1.0` used as a shorthand for float conversion using `numpy`'s broadcasting?

File: examples_apps/run_state_learner.py [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+# exiting the variational circuit. We must use TensorFlow functions to
+# manipulate this data, as well as a ``GradientTape`` to keep track of the
+# corresponding gradients!
+#
+# We choose the following cost function:
+#
+# .. math:: C(\vec{\theta}) = \left| \langle \psi_t \mid U(\vec{\theta})\mid 0\rangle - 1\right|
+#
+# By minimizing this cost function, the variational quantum circuit will
+# prepare a state with high fidelity to the target state.
```
Comment (antalszava) [inline examples_apps/run_state_learner.py:236]: (Minor suggestion): or comparable such that maximizing is included again after the intro
```suggestion
# prepare a state with high fidelity to the target state (hence we achieve maximization).
```

File: examples_apps/run_state_learner.py [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+
+    # Objective function to minimize
+    cost = tf.abs(tf.reduce_sum(tf.math.conj(ket) * target_state) - 1)
+    return cost, fidelity, ket
+
+
+######################################################################
+# Now that the cost function is defined, we can define and run the
+# optimization as well as the quantum engine. Below, we choose the Adam
+# optimizer that is built into TensorFlow, as well as instantiating
```
Comment (antalszava) [inline examples_apps/run_state_learner.py:266]: (Minor suggestion)
```suggestion
# optimizer that is built into TensorFlow, and instantiate
```

File: examples_apps/run_state_learner.py [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+
+    # Stores fidelity at each step
+    fid_progress.append(fid.numpy())
+
+    if fid > best_fid:
+        # store the new best fidelity and best state
+        best_fid = fid.numpy()
+        learnt_state = ket.numpy()
+
+    # one repitition of the optimization
```
Comment (antalszava) [inline examples_apps/run_state_learner.py:296]:
```suggestion
    # one repetition of the optimization
```

File: examples_apps/run_state_learner.py (hunk 6 above)
Comment (antalszava) [reply examples_apps/run_state_learner.py:304]:
```python
Rep: 150 Cost: 0.0955 Fidelity: 0.9607
Rep: 151 Cost: 0.0525 Fidelity: 0.9638
Rep: 152 Cost: 0.0595 Fidelity: 0.9653
Rep: 153 Cost: 0.0867 Fidelity: 0.9638
Rep: 154 Cost: 0.0447 Fidelity: 0.9646
Rep: 155 Cost: 0.0651 Fidelity: 0.9621
Rep: 156 Cost: 0.0905 Fidelity: 0.9603
Rep: 157 Cost: 0.0477 Fidelity: 0.9636
Rep: 158 Cost: 0.0631 Fidelity: 0.9662
Rep: 159 Cost: 0.0893 Fidelity: 0.9657
Rep: 160 Cost: 0.0459 Fidelity: 0.9670
Rep: 161 Cost: 0.0652 Fidelity: 0.9641
Rep: 162 Cost: 0.0920 Fidelity: 0.9615
Rep: 163 Cost: 0.0501 Fidelity: 0.9642
Rep: 164 Cost: 0.0596 Fidelity: 0.9657
Rep: 165 Cost: 0.0859 Fidelity: 0.9643
Rep: 166 Cost: 0.0442 Fidelity: 0.9649
Rep: 167 Cost: 0.0641 Fidelity: 0.9626
Rep: 168 Cost: 0.0893 Fidelity: 0.9610
Rep: 169 Cost: 0.0471 Fidelity: 0.9641
Rep: 170 Cost: 0.0620 Fidelity: 0.9662
Rep: 171 Cost: 0.0880 Fidelity: 0.9655
Rep: 172 Cost: 0.0456 Fidelity: 0.9668
Rep: 173 Cost: 0.0633 Fidelity: 0.9648
Rep: 174 Cost: 0.0894 Fidelity: 0.9627
Rep: 175 Cost: 0.0478 Fidelity: 0.9650
Rep: 176 Cost: 0.0604 Fidelity: 0.9659
Rep: 177 Cost: 0.0863 Fidelity: 0.9645
Rep: 178 Cost: 0.0452 Fidelity: 0.9653
Rep: 179 Cost: 0.0619 Fidelity: 0.9636
Rep: 180 Cost: 0.0870 Fidelity: 0.9620
Rep: 181 Cost: 0.0458 Fidelity: 0.9647
Rep: 182 Cost: 0.0613 Fidelity: 0.9662
Rep: 183 Cost: 0.0867 Fidelity: 0.9656
Rep: 184 Cost: 0.0449 Fidelity: 0.9670
Rep: 185 Cost: 0.0624 Fidelity: 0.9657
Rep: 186 Cost: 0.0880 Fidelity: 0.9640
Rep: 187 Cost: 0.0465 Fidelity: 0.9659
Rep: 188 Cost: 0.0603 Fidelity: 0.9660
Rep: 189 Cost: 0.0861 Fidelity: 0.9644
Rep: 190 Cost: 0.0460 Fidelity: 0.9651
Rep: 191 Cost: 0.0592 Fidelity: 0.9642
Rep: 192 Cost: 0.0839 Fidelity: 0.9631
Rep: 193 Cost: 0.0433 Fidelity: 0.9653
Rep: 194 Cost: 0.0619 Fidelity: 0.9666
Rep: 195 Cost: 0.0864 Fidelity: 0.9662
Rep: 196 Cost: 0.0447 Fidelity: 0.9679
Rep: 197 Cost: 0.0616 Fidelity: 0.9673
Rep: 198 Cost: 0.0873 Fidelity: 0.9655
Rep: 199 Cost: 0.0463 Fidelity: 0.9667
```

File: examples_apps/run_state_learner.py [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+######################################################################
+# We can use the following function to plot the Wigner function of our
+# target and learnt state:
+
+import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
+
+
+def wigner(rho):
+    """This code is a modified version of the 'iterative' method
```
Comment (antalszava) [inline examples_apps/run_state_learner.py:331]: An entry will have to be added to `LICENSE` on this file (and perhaps the license is needed at the beginning of this file as well? :thinking: ).

File: examples_apps/run_state_learner.py [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+---------------
+
+First, we must define the **hyperparameters** of our layer structure:
+
+-  ``cutoff``: the simulation Fock space truncation we will use in the
+   optimization. The TensorFlow backend will perform numerical
+   operations in this truncated Fock space when performing the
+   optimization.
+
+-  ``depth``: The number of layer ansatz in our variational quantum
```
[the remark this answers is not in the document]
Comment (josh146) [reply examples_apps/run_state_learner.py:79]:
```suggestion
-  ``depth``: The number of layers in our variational quantum
```

File: examples_apps/run_state_learner.py (hunk 8 above)
Comment (josh146) [reply examples_apps/run_state_learner.py:92]: I think this is fine, preserve photon number is a pretty common wording

File: examples_apps/run_state_learner.py (hunk 9 above)
Comment (josh146) [reply examples_apps/run_state_learner.py:236]: I'm not sure this sounds entirely right :thinking:

File: examples_apps/run_state_learner.py (hunk 4 above)
Comment (josh146) [reply examples_apps/run_state_learner.py:246]: This is a weird bug/edge case in TensorFlow. While `tf.Tensor` objects _are_ iterables, `tf.Variable` objects aren't!

```python
>>> x = tf.Variable([0.0, 1.0, 2.0])
>>> list(x) # fails since Variables are not iterable
>>> list(x*1.0) # works; multiplying a Variable by a scalar casts it to a Tensor which *is* iterable
[<tf.Tensor: id=17, shape=(), dtype=int32, numpy=0>,
 <tf.Tensor: id=21, shape=(), dtype=int32, numpy=1>,
 <tf.Tensor: id=25, shape=(), dtype=int32, numpy=2>]
```

File: examples_apps/run_state_learner.py (hunk 6 above)
Comment (josh146) [reply examples_apps/run_state_learner.py:304]: Weird @trbromley, it trains fine for me locally (and matches @antalszava's output!) I wonder why it's not working for you 🤔 What version of TensorFlow do you have installed?

File: examples_apps/run_state_learner.py (hunk 12 above)
Comment (josh146) [reply examples_apps/run_state_learner.py:331]: > and perhaps the license is needed at the beginning of this file as well?

We don't typically worry about adding licenses to each file in the documentation, as there are a lot of files here that don't allow commenting! This is fine, since the license holds for the entire repository.

> An entry will have to be added to LICENSE on this file 

I wouldn't worry about this for now, since all tutorials are due to be moved to a separate repository soon.

File: doc/introduction/tutorials.rst [hunk 14]
```diff
@@ -124,7 +123,7 @@ Optimization and machine learning
 
     /tutorials/tutorial_machine_learning
     /gallery/minimizing_correlations/minimizing_correlations
-    /gallery/state_learner/StateLearning
+    ../tutorials_apps/run_state_learner
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/tutorials.rst:126]: Since we're moving the tutorials out of this repo soon, I wouldn't worry about it :) We can build up the new structure from scratch in the applications repo.

File: doc/conf.py [hunk 15]
```diff
@@ -24,16 +24,6 @@
...
 
-class Mock(MagicMock):
-    __name__ = 'foo'
-
-    @classmethod
-    def __getattr__(cls, name):
-        return MagicMock()
-
-
-sys.modules["tensorflow"] = Mock(__version__="1.3")
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/conf.py:35]: No, this needed to be removed so that Sphinx can import TensorFlow, in order to execute the tutorial :)

**Comment by josh146** [comment]
> I'm also worried if this will finally stop us being able to build on RTD 🤔 Not because this tutorial is particularly long (it took me 2 mins locally), but we just seem close!

Let's see how it goes 😆 I've got my fingers crossed, because we're not close to hitting the timeout on read the docs (instead I'm more worried about memory usage). I'm beginning work on the new photonics-applications repo, so we should be able to move the tutorials very soon.

File: examples_apps/run_state_learner.py (hunk 4 above)
Comment (josh146) [reply examples_apps/run_state_learner.py]: @trbromley: I've updated this line to make it much simpler, I realised you don't need a nested dictionary comprehension if you just flatten both lists first! Check it out and let me know what you think.

Comment (trbromley) [reply examples_apps/run_state_learner.py]: Great, looks much better thanks!

File: examples_apps/run_state_learner.py (hunk 6 above)
Comment (trbromley) [reply examples_apps/run_state_learner.py:325]: :heavy_check_mark: Using TensorFlow==2.0.0:
```python
Rep: 195 Cost: 0.0212 Fidelity: 0.9909
Rep: 196 Cost: 0.0236 Fidelity: 0.9910
Rep: 197 Cost: 0.0212 Fidelity: 0.9910
Rep: 198 Cost: 0.0235 Fidelity: 0.9911
Rep: 199 Cost: 0.0211 Fidelity: 0.9910
```
:x: Using TensorFlow==2.1.0:
```python
Rep: 195 Cost: 1.0138 Fidelity: 0.0002
Rep: 196 Cost: 1.0133 Fidelity: 0.0002
Rep: 197 Cost: 1.0128 Fidelity: 0.0002
Rep: 198 Cost: 1.0126 Fidelity: 0.0002
Rep: 199 Cost: 1.0124 Fidelity: 0.0002
```
Interesting how the one that works doesn't match yours and @antalszava's results.

Comment (josh146) [reply examples_apps/run_state_learner.py:325]: That's super strange...

Comment (trbromley) [reply examples_apps/run_state_learner.py:325]: Maybe it's something specific to my setup, do you get a problem with 2.1.0? And if so, isn't this something to worry about :sweat_smile:?

Comment (josh146) [reply examples_apps/run_state_learner.py:325]: @trbromley, can confirm that it doesn't train for me either when I upgrade to 2.1.0, but as soon as I downgrade to 2.0 it works again 🤔 I have no idea why this is the case

Comment (antalszava) [reply examples_apps/run_state_learner.py:325]: Same for me as well with `2.1`.

File: examples_apps/run_state_learner.py (hunk 4 above)
Comment (antalszava) [reply examples_apps/run_state_learner.py]: Great, thank you! With the new version `tf.reshape(weights, [-1]))` would do the casting, right?

It might be worth capturing it in a comment: `# As TensorFlow Variables are not iterable, they are being cast to Tensors through reshaping`

File: examples_apps/run_state_learner.py [hunk 16]
```diff
@@ -0,0 +1,453 @@
...
+r2 = tf.random.normal(shape=[depth], stddev=passive_sd)
+
+# kerr gate
+kappa = tf.random.normal(shape=[depth], stddev=active_sd)
+
+
+######################################################################
+# For convenience, we store the TensorFlow variables representing the
+# weights as a tensor, as well as creating the corresponding
+# single-mode Strawberry Fields program and parameters:
```
Comment (antalszava) [inline examples_apps/run_state_learner.py:150]: This part of the comment refers to the next section now (the code parts were moved if I'm not mistaken)
```suggestion
# weights as a tensor:
```

File: examples_apps/run_state_learner.py [hunk 17]
```diff
@@ -0,0 +1,453 @@
...
+
+
+#######################################################################
+# Now that the cost function is defined, we can define and run the
+# optimization. Below, we choose the Adam
+# optimizer that is built into TensorFlow, and instantiate
+# the Strawberry Fields TensorFlow backend.
+
+opt = tf.keras.optimizers.Adam(learning_rate=lr)
+eng = sf.Engine("tf", backend_options={"cutoff_dim": cutoff})
```
Comment (antalszava) [inline examples_apps/run_state_learner.py:292]: `eng` is used in the cost function, would it make sense to define it earlier?

File: examples_apps/run_state_learner.py (hunk 16 above)
Comment (josh146) [reply examples_apps/run_state_learner.py:150]: Good catch, I'll move it earlier!

File: examples_apps/run_state_learner.py (hunk 17 above)
Comment (josh146) [reply examples_apps/run_state_learner.py:292]: Good catch, I'll move it earlier!
